### ISSUE TITLE
virtual query overflow?

### ISSUE THREAD

**Issue by kitty8904** [cheat-engine/cheat-engine#954]
<img width="1009" alt="virtualquery" src="https://user-images.githubusercontent.com/18622105/67632293-f3eae400-f8e4-11e9-8781-76bf3f45af79.png">

while (Virtualqueryex(processhandle,pointer(currentBaseAddress),mbi,sizeof(mbi))<>0) and (currentBaseAddress<stopaddress) and ((currentBaseAddress+mbi.RegionSize)>currentBaseAddress) do
//last check is done to see if it wasn't a 64-bit overflow.

Self=0x1912f0
CURRENTBASEADDRESS=18446740225418854400
MBI={BASEADDRESS = 0xfffffc8000000000, ALLOCATIONBASE = 0xfffffc8000000000, ALLOCATIONPROTECT = 1, REGIONSIZE = 3848290697216, STATE = 65536, PROTECT = 1, _TYPE = 131072}
I=0
J=0
BLOCKSIZE=0
CURRENTBLOCKSIZE=0
TOTALPROCESSMEMORYSIZE=18446740225418854400
LEFTFROMPREVIOUS=0
OFFSETINCURRENTREGION=4295476230
ISWRITABLE=true
ISEXECUTABLE=true
ISCOPYONWRITE=false
VALIDREGION=true
DATATYPE=
VQECACHEFLAG=5

baseaddress:: 0xfffffc8000000000 + ?? -> Overflow?

Lazarus debugging due to kernel memory area checking

Overflow occurs in memscan.pas

Where do you check for overflow?

Does BSOD occur because of overflow?

**Comment by cheat-engine** [comment]
No, if there is an overflow than the system still wouldn't BSOD, ce would just hang.
BSOD happens because likely the page being read for the page information got swapped out and replaced by data messing up the read.

Also, what is the BSOD you get? Exception in nonpaged area? Clock timeout? IRQL not equal?

**Comment by kitty8904** [comment]
Searching for virtualquery in kernel memory area

BSOD occurs in a section you can't read somewhere

And when debugging Lazarus

Kernel address value overflow occurs in virtualquery loop statement

**Comment by kitty8904** [comment]
#if (NTDDI_VERSION >= NTDDI_WIN2K)
//@[comment("MVI_tracked")]
_IRQL_requires_max_(PASSIVE_LEVEL)
_Must_inspect_result_
NTSYSAPI
NTSTATUS
NTAPI
ZwQueryVirtualMemory(
    _In_ HANDLE ProcessHandle,
    _In_opt_ PVOID BaseAddress,
    _In_ MEMORY_INFORMATION_CLASS MemoryInformationClass,
    _Out_writes_bytes_(MemoryInformationLength) PVOID MemoryInformation,
    _In_ SIZE_T MemoryInformationLength,
    _Out_opt_ PSIZE_T ReturnLength
    );
#endif

Doesn't the cheat engine use ZwQueryVirtualMemory?

Windows ZwQueryVirtualMemory has problems, so do you develop and retrieve functions?

**Comment by cheat-engine** [comment]
That requires a handle

**Comment by kitty8904** [comment]
In the user mode query possible?
KeAttachProcess
ZwQueryVirtualMemory (ZwCurrentProcess
KeDetachProcess

Is this because of kernel mode search?

**Comment by cheat-engine** [comment]
Might work, give it a try
